gene/insert.go

```diff
@@ -0,0 +1,326 @@
+	}
+
+	if genomeIndexPos > len(g.genomeSeq)-1 {
+		return -1, ErrInputPosNotInGene
+	}
+
+	// If no error, then append changeLog
+	g.changeLog = append(g.changeLog, log)
+
+	return genomeIndexPos, nil
+}
+
+// insertAdjust is a wrapper for insertStable that adjusts the start and end positions of UTRs and CDS.
+func insertAdjust(g *Gene, destSeq *[]dna.Base, insPos int, insSeq []dna.Base) error {
+	var err error
+	err = insertStable(destSeq, insPos, insSeq)
+	if errors.Is(err, ErrExceedCapNewSliceCreated) {
+		err = nil
+	}
+
+	if g.utrFive.start > insPos {
+		g.utrFive.start += len(insSeq)
+	}
+
+	if g.utrFive.end > insPos {
+		g.utrFive.end += len(insSeq)
+	}
+
+	if g.utrThree.start > insPos {
+		g.utrThree.start += len(insSeq)
+	}
+
+	if g.utrThree.end > insPos {
+		g.utrThree.end += len(insSeq)
+	}
+
+	if g.codingSeq.start > insPos {
+		g.codingSeq.start += len(insSeq)
+	}
+
+	if g.codingSeq.end > insPos {
+		g.codingSeq.end += len(insSeq)
+	}
+
+	g.utrFive.seq = g.cdnaSeq[g.utrFive.start:g.utrFive.end]
+	g.utrThree.seq = g.cdnaSeq[g.utrThree.start:g.utrThree.end]
+	g.codingSeq.seq = g.cdnaSeq[g.codingSeq.start:g.codingSeq.end]
+
+	return nil
+}
+
+// insertStable is identical to Insert, but conserves the
+// underlying memory of the input destSeq slice if possible
+func insertStable(destSeq *[]dna.Base, insPos int, insSeq []dna.Base) error {
+	if insPos < 0 || insPos > len(*destSeq) {
+		return errInvalidPosition
+	}
+	if len(*destSeq)+len(insSeq) > cap(*destSeq) {
+		*destSeq = dna.Insert(*destSeq, insPos, insSeq)
+		return ErrExceedCapNewSliceCreated
+	}
+	*destSeq = (*destSeq)[:len(*destSeq)+len(insSeq)]                                    // extend the LEN of the slice to include the inserted bases
+	copy((*destSeq)[insPos+len(insSeq):cap(*destSeq)], (*destSeq)[insPos:cap(*destSeq)]) // shift bases in slice by len(insSeq) at the position of the insertion
+	copy((*destSeq)[insPos:insPos+len(insSeq)], insSeq)                                  // add in the inserted bases
+	return nil
+}
+
+// frameshiftTranslate translates the result of a frameshift going into the 3'UTR sequence if necessary.
+// Returns a stop-terminated sequence of amino acids
+func frameshiftTranslate(shiftedCds []dna.Base, utrThree []dna.Base) []dna.AminoAcid {
+	answer := make([]dna.AminoAcid, 0, (len(shiftedCds)+len(utrThree))/3)
+
+	frameOffset := len(shiftedCds) % 3
+
+	for i := 0; i < len(shiftedCds)-frameOffset; i += 3 {
+		if i+3 > len(shiftedCds) {
+			break
+		}
+		answer = append(answer, dna.TranslateSeq(shiftedCds[i : i+3])[0])
+		if answer[len(answer)-1] == dna.Stop {
+			return answer
+		}
+	}
+
+	transitionCodon := append(shiftedCds[len(shiftedCds)-frameOffset:], utrThree[:3-frameOffset]...)
+	answer = append(answer, dna.TranslateSeq(transitionCodon)[0])
+	if answer[len(answer)-1] == dna.Stop {
+		return answer
+	}
+
+	remainingUtrThree := utrThree[3-frameOffset:]
+
+	for i := 0; i < len(remainingUtrThree); i += 3 {
+		if i+3 > len(remainingUtrThree) {
```

Comment:
Should this be in the for loop condition instead of here?

Reply:
Yes